feature core/input: validation in input components

1307-Dev opened the pull request:
## 💡 What is the current behavior?

There are some errors which prevent validations from happening has expected.

GitHub Issue Number: #1799 
JIRA: [IX-2595](https://agileworld.siemens.cloud/jira/browse/IX-2595)

## 🆕 What is the new behavior?

The acceptance criteria for this ticket has been met- 
Required input:

Invalid input > Removing value with keyboard > Stays invalid
Invalid input > Remove touched state (e.g. clear button) > Valid again
Invalid input > Programmatically setting it to empty > Stays invalid
Valid input > Removing value with keyboard > It is invalid
Valid input > Remove touched state (e.g. clear button) > Valid
Valid input > Programmatically setting it to empty > It is invalid
Not required input:

Invalid input > Removing value with keyboard > Valid
Invalid input > Remove touched state (e.g. clear button) > Valid again
Invalid input > Programmatically setting it to empty > Valid
Valid input > Removing value with keyboard > Valid
Valid input > Remove touched state (e.g. clear button) > Valid
Valid input > Programmatically setting it to empty > Valid


Code duplication has been reduced for date and time input.
Code refactoring had to do for date and time input.
Common clear functionality is added to support all the inputs. 
Slot implementation simplified and reduced duplication in file [packages/core/src/components/input/input.fc.tsx‎]

## 🏁 Checklist

A pull request can only be merged if all of these conditions are met (where applicable):

- [ ] 🦮 Accessibility (a11y) features were implemented
- [ ] 🗺️ Internationalization (i18n) - no hard coded strings
- [ ] 📲 Responsiveness - components handle viewport changes and content overflow gracefully
- [ ] 📕 Add or update a Storybook story
- [ ] 📄 Documentation was reviewed/updated [siemens/ix-docs](https://github.com/siemens/ix-docs)
- [x] 🧪 Unit tests were added/updated and pass (`pnpm test`)
- [ ] 📸 Visual regression tests were added/updated and pass ([Guide](https://github.com/siemens/ix/blob/main/CONTRIBUTING.md#visual-regression-testing))
- [x] 🧐 Static code analysis passes (`pnpm lint`)
- [x] 🏗️ Successful compilation (`pnpm build`, changes pushed)

## 👨‍💻 Help & support

1307-Dev committed:
Pushed changes for input, textarea, number input and select

khathija-ahamadi committed:
Checkbox - form validations handled.

khathija-ahamadi committed:
Merge branch '(feature/core)validation-input-components' of https://github.com/1307-Dev/ix into (feature/core)validation-input-components

khathija-ahamadi committed:
Checkbox unit tests added for noValidate scenraios.

1307-Dev committed:
removed reset for select component.

1307-Dev committed:
Merge branch '(feature/core)validation-input-components' of https://github.com/1307-Dev/ix into (feature/core)validation-input-components

khathija-ahamadi committed:
Lint and prettier fixes - Checkbox component.

khathija-ahamadi committed:
Merge branch '(feature/core)validation-input-components' of https://github.com/1307-Dev/ix into (feature/core)validation-input-components

1307-Dev committed:
Added lint fixes

khathija-ahamadi committed:
Checkbox component refactored - congnitive complexity fixed (sonar)

khathija-ahamadi committed:
Merge branch '(feature/core)validation-input-components' of https://github.com/1307-Dev/ix into (feature/core)validation-input-components

khathija-ahamadi committed:
Radio form validations for noValidate cases implemented.

khathija-ahamadi committed:
Checkbox and radio component refactored.

khathija-ahamadi committed:
Radio button component refactored.

khathija-ahamadi committed:
Duplicate issue of sonar qube for checkbox component fixed.

khathija-ahamadi committed:
Sonar fixes

khathija-ahamadi committed:
Sonar fixes.

khathija-ahamadi committed:
Checkbox and radio sonar fixes.

khathija-ahamadi committed:
Checkbox and radio component refactored.

khathija-ahamadi committed:
Radio and checkbox sonar fixes.

1307-Dev committed:
Merge branch 'main' into (feature/core)validation-input-components

khathija-ahamadi committed:
Radio group linting issue fixed.

khathija-ahamadi committed:
Merge branch '(feature/core)validation-input-components' of https://github.com/1307-Dev/ix into (feature/core)validation-input-components

RamVinayMandal committed:
Merge branch 'siemens:main' into (feature/core)validation-input-components

RamVinayMandal committed:
Date and Time input implementation

1307-Dev committed:
Renamed method to clear and modified the functionality

RamVinayMandal committed:
Merge branch 'main' into (feature/core)validation-input-components

khathija-ahamadi committed:
Fix unit test failing due to infra issue.

khathija-ahamadi committed:
Merge branch '(feature/core)validation-input-components' of https://github.com/1307-Dev/ix into (feature/core)validation-input-components

khathija-ahamadi committed:
Fix unit tests failing due to infra issue.

khathija-ahamadi committed:
Fix: First appy validation classes and then render and then assert if the classes are present - checkbox.

khathija-ahamadi committed:
wait for timeout added.

1307-Dev committed:
Modified failing test cases

RamVinayMandal committed:
Fixed sonarcube issues

RamVinayMandal committed:
Merge branch '(feature/core)validation-input-components' of github.com:1307-Dev/ix into (feature/core)validation-input-components

RamVinayMandal committed:
Lint issue fixed

khathija-ahamadi committed:
Revert unit tests for checkbox and radio component.

khathija-ahamadi committed:
Merge branch '(feature/core)validation-input-components' of https://github.com/1307-Dev/ix into (feature/core)validation-input-components

1307-Dev committed:
Added clear method for radiobutton, checkbox and their group components.

RamVinayMandal committed:
Merge branch 'main' into (feature/core)validation-input-components

RamVinayMandal committed:
Resolved conflicts

RamVinayMandal committed:
Merge branch 'siemens:main' into (feature/core)validation-input-components

nuke-ellington reviewed (commented):
Before thorough review can be conducted please:

1. Resolve conflicts
2. Reduce code duplication (Sonarqube)
3. Fix pipeline (unit test stage)

RamVinayMandal committed:
Merge branch 'main' into (feature/core)validation-input-components

RamVinayMandal committed:
feat(core/picker-input): refactor date and time input components to extend BasePickerInput for shared functionality

RamVinayMandal committed:
refactor(core): streamline picker input components and improve method naming

RamVinayMandal committed:
refactor(core/input): enhance input clearing logic and improve state management

RamVinayMandal committed:
Merge branch 'main' into (feature/core)validation-input-components

RamVinayMandal committed:
feat(core/date-input, time-input): add isInputInvalid state and improve type annotations in tests

RamVinayMandal committed:
Merge branch '(feature/core)validation-input-components' of github.com:1307-Dev/ix into (feature/core)validation-input-components

RamVinayMandal committed:
Merge branch 'main' into (feature/core)validation-input-components

RamVinayMandal committed:
feat(core): enhance input validation and state management in input components

RamVinayMandal committed:
Merge branch '(feature/core)validation-input-components' of github.com:1307-Dev/ix into (feature/core)validation-input-components

RamVinayMandal committed:
feat: enhance validation and input handling in date and time components

- Refactored validation logic in DateInput and TimeInput components to utilize new picker methods.
- Introduced context-based picker methods for better state management and validation handling.
- Removed redundant code and improved the structure of input methods for clarity and maintainability.
- Updated event handling to streamline input validation and dropdown interactions.
- Enhanced the common input utilities to support new validation and event handling patterns.

RamVinayMandal committed:
refactor: enhance base picker input with new utility functions and state management

- Introduced new utility functions for managing dropdown visibility and input validation.
- Refactored state management in BasePickerInput to improve reactivity and maintainability.
- Added support for class mutation observers to handle dynamic class changes.
- Updated picker input methods to streamline event handling and validation synchronization.
- Created a new types file for picker input to centralize type definitions and improve code clarity.
- Simplified input rendering logic and improved integration with dropdown components.

RamVinayMandal committed:
refactor: rename build functions to improve clarity in picker input methods

RamVinayMandal committed:
feat: update dropdown test IDs in date and time input components for consistency

RamVinayMandal committed:
feat: enhance validity state tracking and input validation for date and time components

RamVinayMandal committed:
fix: update time picker interaction to use icon button for opening dropdown

RamVinayMandal committed:
Merge branch '(feature/core)validation-input-components' of github.com:1307-Dev/ix into (feature/core)validation-input-components

RamVinayMandal committed:
test: add validation for initial date input value in form

RamVinayMandal committed:
fix: update SPDX copyright year to 2026 in picker input files

khathija-ahamadi committed:
Refactored radio and checkbox component.

khathija-ahamadi committed:
Merge branch '(feature/core)validation-input-components' of https://github.com/1307-Dev/ix into (feature/core)validation-input-components

khathija-ahamadi committed:
Lint fix radio group component.

varun-srinivasa committed:
Merge branch 'main' into (feature/core)validation-input-components

RamVinayMandal committed:
Merge remote-tracking branch 'origin/main' into (feature/core)validation-input-components

varun-srinivasa committed:
Merge branch '(feature/core)validation-input-components' of https://github.com/1307-Dev/ix into (feature/core)validation-input-components

varun-srinivasa committed:
updated failing scenario - dynamic item not focusable

khathija-ahamadi committed:
Update assert condition in radio group.

khathija-ahamadi committed:
Assert updated.

khathija-ahamadi committed:
Radio test updated.

khathija-ahamadi committed:
Radio changes.

khathija-ahamadi committed:
Removed comment.

RamVinayMandal committed:
Fixed placement of datatypes and info which got inversely placed while fixing merge conflicts

RamVinayMandal committed:
Merge branch '(feature/core)validation-input-components' of github.com:1307-Dev/ix into (feature/core)validation-input-components

RamVinayMandal committed:
Aligned changes with main to omit unnecessary changes

1307-Dev committed:
Merge branch 'main' into (feature/core)validation-input-components

RamVinayMandal committed:
Pushed auto generated files after build

1307-Dev committed:
Reverted some preview examples

nuke-ellington commented on packages/core/src/components/utils/input/picker-input.types.ts:
There is a bunch of redundancy in these interfaces (e.g. touched). Is this really neccessary?

khathija-ahamadi committed:
Refactored radio and checkbox component.

khathija-ahamadi committed:
Radio component updated.

khathija-ahamadi committed:
Gemini review comments handled for radio and checkbox..

khathija-ahamadi commented:
/gemini review

khathija-ahamadi committed:
Refactored checkbox and radio component.

khathija-ahamadi commented on packages/core/src/components/checkbox-group/checkbox-group.tsx:
```diff
@@ -144,9 +175,127 @@ export class CheckboxGroup
+            touched?: boolean;
+            formSubmissionAttempted?: boolean;
+          }
+        ).formSubmissionAttempted
```
Refactored as per the suggestion.

khathija-ahamadi commented:
/gemini review

varun-srinivasa committed:
Merge branch 'main' of https://github.com/1307-Dev/ix into (feature/core)validation-input-components

khathija-ahamadi committed:
Refactored utils component.

khathija-ahamadi committed:
Merge branch '(feature/core)validation-input-components' of https://github.com/1307-Dev/ix into (feature/core)validation-input-components

RamVinayMandal committed:
Addressed Gemini code review - Refactor validation handling in input components and improve state synchronization

RamVinayMandal commented on packages/core/src/components/time-input/time-input.tsx:
```diff
@@ -319,35 +307,35 @@ export class TimeInput implements IxInputFieldComponent<string> {
+  @HookValidationLifecycle()
+  hookValidationLifecycle(results: ValidationResults) {
+    this.pickerMethods.hookValidationLifecycle(results);
+  }
```
Addressed.

RamVinayMandal commented on packages/core/src/components/time-input/time-input.tsx:
```diff
@@ -319,35 +307,35 @@ export class TimeInput implements IxInputFieldComponent<string> {
     if (this.isInputInvalid) {
       this.time = null;
     } else {
       this.watchValue();
```
Addressed

RamVinayMandal committed:
Refactor picker input methods and types for improved type safety and clarity. Reduced boilerplate code.

RamVinayMandal commented on packages/core/src/components/select/select.tsx:
```diff
@@ -525,6 +573,45 @@ export class Select implements IxInputFieldComponent<string | string[]> {
+        capture: true,
+      });
+    }
+  }
```
The public `clear()` method is not needed for `ix-select` as discussed with Andreas.
